src/neo-vm/ReferenceCounter.cs

```diff
@@ -0,0 +1,104 @@
+using Neo.VM.Types;
+using System.Collections.Generic;
+
+namespace Neo.VM
+{
+    public sealed class ReferenceCounter
+    {
+        private class Entry
+        {
+            public int StackReferences;
+            public Dictionary<CompoundType, int> ObjectReferences;
+        }
+
+        private readonly Dictionary<CompoundType, Entry> counter = new Dictionary<CompoundType, Entry>(ReferenceEqualityComparer.Default);
+        private readonly List<CompoundType> zero_referred = new List<CompoundType>();
+        private int references_count = 0;
+
+        public int Count => references_count;
+
+        internal void AddReference(StackItem referred, CompoundType parent)
+        {
+            references_count++;
+            if (!(referred is CompoundType compound)) return;
+            if (!counter.TryGetValue(compound, out Entry tracing))
+            {
+                tracing = new Entry();
+                counter.Add(compound, tracing);
+            }
+            if (tracing.ObjectReferences is null)
+                tracing.ObjectReferences = new Dictionary<CompoundType, int>(ReferenceEqualityComparer.Default);
+            if (tracing.ObjectReferences.TryGetValue(parent, out int count))
+                count++;
+            else
+                count = 1;
+            tracing.ObjectReferences[parent] = count;
+        }
+
+        internal void AddStackReference(StackItem referred)
+        {
+            references_count++;
+            if (!(referred is CompoundType compound)) return;
+            if (counter.TryGetValue(compound, out Entry entry))
+                entry.StackReferences++;
+            else
+                counter.Add(compound, new Entry { StackReferences = 1 });
+        }
+
+        internal void CheckZeroReferred()
+        {
+            if (zero_referred.Count == 0) return;
+            HashSet<CompoundType> toBeDestroyed = new HashSet<CompoundType>(ReferenceEqualityComparer.Default);
+            foreach (CompoundType compound in zero_referred)
+            {
+                HashSet<CompoundType> toBeDestroyedInLoop = new HashSet<CompoundType>(ReferenceEqualityComparer.Default);
+                Queue<CompoundType> toBeChecked = new Queue<CompoundType>();
```

Comment:
```suggestion
                Queue<CompoundType> toBeChecked = new Queue<CompoundType>(1);
```

Reply:
If the queue will have more the one item, we will need a resizing on the queue.

Reply:
```suggestion
                Queue<CompoundType> toBeChecked = new Queue<CompoundType>(4);
```


![image](https://user-images.githubusercontent.com/3167973/69866888-97913080-12a5-11ea-9f81-7cb9e84c1ca0.png)

I think that it should start with 4. Because it will be the first resize that it will do everytime.

Reply:
I think we'd better not change the initial capacity unless we can predict the size it may reach. Different CLR may have different implementations for the queue resizing.

Reply:
I got you, but then... 1 is your number! :P because always you put one element at the next line, so you can save one resize ever

Reply:
https://github.com/dotnet/corefx/blob/e99ec129cfd594d53f4390bf97d1d736cff6f860/src/System.Collections/src/System/Collections/Generic/Queue.cs#L184-L200

If you don't specify an initial size, it will allocate array space on the first `Enqueue()`. Otherwise it will be allocated when constructing the `Queue`. It looks the same to me.

Reply:
But you will call first https://github.com/dotnet/corefx/blob/e99ec129cfd594d53f4390bf97d1d736cff6f860/src/System.Collections/src/System/Collections/Generic/Queue.cs#L42 for allocate nothing.

The difference is one allocation less

Reply:
No, no, no...

https://github.com/dotnet/corefx/blob/e99ec129cfd594d53f4390bf97d1d736cff6f860/src/Common/src/CoreLib/System/Array.cs#L682-L685

`Array.Empty<T>` doesn't allocate.

Reply:
You are right :)